packages/workbox-precaching/src/utils/PrecacheStrategy.ts

```diff
@@ -0,0 +1,138 @@
+/*
+  Copyright 2020 Google LLC
+
+  Use of this source code is governed by an MIT-style
+  license that can be found in the LICENSE file or at
+  https://opensource.org/licenses/MIT.
+*/
+
+import {copyResponse} from 'workbox-core/copyResponse.js';
+import {getFriendlyURL} from 'workbox-core/_private/getFriendlyURL.js';
+import {logger} from 'workbox-core/_private/logger.js';
+import {WorkboxError} from 'workbox-core/_private/WorkboxError.js';
+import {WorkboxPluginCallbackParam} from 'workbox-core/types.js';
+import {Strategy, StrategyOptions} from 'workbox-strategies/Strategy.js';
+import {StrategyHandler} from 'workbox-strategies/StrategyHandler.js';
+
+import '../_version.js';
+
+
+async function copyRedirectedResponses({response}: WorkboxPluginCallbackParam['cacheWillUpdate']) {
+  return response.redirected ? await copyResponse(response) : response;
+}
+
+
+class PrecacheStrategy extends Strategy {
+  constructor(options: StrategyOptions) {
+    super(options);
+
+    // Redirected responses cannot be used to satisfy a navigation request, so
+    // any redirected response must be "copied" rather than cloned, so the new
+    // response doesn't contain the `redirected` flag. See:
+    // https://bugs.chromium.org/p/chromium/issues/detail?id=669363&desc=2#c1
+    this.plugins.push({cacheWillUpdate: copyRedirectedResponses});
+  }
+
+  _handle(request: Request, handler: StrategyHandler) {
+    // If this is an `install` event then populate the cache. If this is a
+    // `fetch` event (or any other event) then respond with the cached response.
+    if (handler.event && handler.event.type === 'install') {
+      return this._handleInstall(request, handler);
+    }
+    return this._handleFetch(request, handler);
+  }
+
+  async _handleFetch(request: Request, handler: StrategyHandler) {
+    let response = await handler.cacheMatch(request);
+
+    if (!response) {
+      // Fall back to the network if we don't have a cached response
+      // (perhaps due to manual cache cleanup).
+      if (handler.params &&
+          handler.params.fallbackToNetwork === false) {
```

Comment:
When would this code get called with `handler.params.fallbackToNetwork === false`?

Also, we can start introducing the optional chaining operator into the codebase, since we run everything though `preset-env` anyway.

Reply:
> When would this code get called with `handler.params.fallbackToNetwork === false`?

That happens here: https://github.com/GoogleChrome/workbox/blob/e4b8c673ec172767592aca46ac1683ab38184666/packages/workbox-precaching/src/PrecacheController.ts#L479.

> Also, we can start introducing the optional chaining operator into the codebase, since we run everything though `preset-env` anyway.

I'm not sure about that. We do use babel for our production builds, but for anyone who uses their own bundler, they'd just get our source code. We could change that, but unless we do, I don't want us to use JS features that require specific transpilation targets that would be easy for developers to miss.

Reply:
Okay, no optional chaining then.